### ISSUE TITLE
ORA-38104 when using ON DUPLICATE KEY UPDATE in Oracle

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8158]
Since jOOQ 3.11 (#6462) the `ON DUPLICATE KEY UPDATE` clause is emulated in other databases by running a `MERGE` statement that checks for duplicates in all of a table's unique keys, not just the primary key. This causes ORA-38104 errors in the presence of unique keys. For example, this query from the integration tests doesn't work:

```sql
merge into "TEST"."T_AUTHOR"
using (select 1 "one" from dual)
on ("TEST"."T_AUTHOR"."ID" = 1 or "TEST"."T_AUTHOR"."LAST_NAME" = 'Hesse')
when matched then update set "TEST"."T_AUTHOR"."LAST_NAME" = 'Hesse'
when not matched then insert ("ID", "LAST_NAME")
values (1, 'Hesse')
```

But we can apply a trick and transform the query into this:

```sql
merge into (select ID, nvl(LAST_NAME, LAST_NAME) as LAST_NAME1, LAST_NAME from T_AUTHOR)
using (select 1 one from dual)
on (ID = 1 or LAST_NAME1= 'Hesse')
when matched then update set LAST_NAME = 'Hesse'
when not matched then insert (ID, LAST_NAME)
values (1, 'Hesse');
```

An alternative would be to generate this query:

```sql
merge into T_AUTHOR
using (select 1 one from dual)
on (ID = 1 or (select LAST_NAME from dual) = 'Hesse')
when matched then update set LAST_NAME = 'Hesse'
when not matched then insert (ID, LAST_NAME)
values (1, 'Hesse');
```

But the execution plan shows that this prevents any index usage:

![image](https://user-images.githubusercontent.com/734593/50597572-1d470900-0ea8-11e9-9a9b-afdc6678e487.png)

As opposed to the acceptable (but more difficult to emulate) `NVL()` solution:

![image](https://user-images.githubusercontent.com/734593/50597590-2d5ee880-0ea8-11e9-8d93-a079bce648f0.png)

This also seems to work, without any penalty with respect to the execution plan:

```sql
merge into T_AUTHOR
using (select 1 one from dual)
on (ID = 1 or (last_name, 'dummy') = (('Hesse', 'dummy')))
when matched then update set LAST_NAME = 'Hesse'
when not matched then insert (ID, LAST_NAME)
values (1, 'Hesse');
```

----
See also https://blog.jooq.org/2019/01/02/how-to-work-around-ora-38104-columns-referenced-in-the-on-clause-cannot-be-updated/

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.12. Backport scheduled for 3.11.9 (#8159). 3.10 is not affected

**Comment by lukaseder** [comment]
These workarounds will be superseded by https://github.com/jOOQ/jOOQ/issues/11205
